### PR TITLE
Add E2E support for THD format

### PR DESCRIPTION
**Pull request by xiaoyao0115**
## Description

This PR adds **Sequence Packing (THD format)** E2E support to MCore. Dev branch PR:https://github.com/NVIDIA/Megatron-LM/pull/2924

The core missing functionalities of THD in MCore are:
* data iterator cannot handle THD meta data, like cu_seqlens, max_seqlens.
* num_microbatches is fixed.
* PackParams are not passing between PP ranks.

#### 1. Add a data_iterator wrapper (megatron/core/datasets/data_schedule.py::wrap_dataloader)

A wrapper function that intercepts the data iterator to perform scheduling and packing:

- **Schedule & Pack**: Extracts data from the data iterator, schedules sequences across DP×CP ranks, and packs them into microbatches with `cu_seqlens` metadata.
- **Returns packing results**: Returns the packed `num_microbatches` along with two parameters for FLOPs calculation: `num_total_tokens_this_global_batch` and `sequence_square_sum_this_global_batch`.
- **TP broadcast**: Broadcasts `num_microbatches` and FLOPs parameters across TP ranks since only TP rank 0 has access to the data iterator.
- **PP broadcast**: When using PP, middle PP stages (not first or last) require metadata (`cu_seqlens`, `cu_seqlens_padded`, `max_seqlen`, etc.) to be broadcast from PP rank 0 for correct computation.

#### 2. Mock SFT Dataset Support

Supports mock datasets for testing and benchmarking with configurable sequence length distributions.
There are two modes of mock sft dataset:
- **File mode**: Load sequence lengths from an external CSV, example json:
  ```json
  {"mode": "file", "path": "/path/to/seqlens.csv"}
  ```
- **Distribution mode**: Generate sequence lengths from a distribution (currently supports lognormal), example json:
  ```json
  {"mode": "distribution", "type": "lognormal", "min_seq_len": 1024, "max_seq_len": 8192, "mean_seq_len": 4096, "lognormal_sigma": 1.1}
  ```

#### Before vs After

```mermaid
graph LR
    subgraph Before
        A1[DataIterator] --> B1[get_batch]
        B1 --> C1[forward_backward]
        C1 --> D1[Fixed seq_len FLOPs]
    end
    subgraph After
        A2[DataIterator] --> W[wrap_dataloader]
        W -->|schedule + pack| B2[PackedDataIterator]
        W -->|broadcast| M[num_microbatches + flops_params]
        B2 --> C2[get_batch_for_sequence_packing]
        C2 --> D2[forward_backward]
        D2 --> E2[Dynamic FLOPs]
        M   --> E2
    end
```

#### Execution Flow

```mermaid
sequenceDiagram
    participant Train as training.py
    participant Schedule as schedules.py
    participant Wrap as wrap_iterator_helper
    participant DataSched as data_schedule.py
    participant GetBatch as get_batch_for_seq_packing

    Train->>Schedule: forward_backward_*(data_iterator)
    Schedule->>Wrap: wrap_iterator_helper(config, data_iterator)
    Wrap->>DataSched: wrap_dataloader(data_iterator, scheduler_type)
    
    Note over DataSched: 1. Gather global seqlens across DP
    Note over DataSched: 2. Scheduler assigns sequences to microbatches
    Note over DataSched: 3. All-to-all redistribute samples
    Note over DataSched: 4. Pack into microbatches
    Note over DataSched: 5. Broadcast to TP/PP ranks
    
    DataSched-->>Schedule: (packed_iter, num_mbs, total_tokens, seq_sq_sum)
    
    loop for each microbatch
        Schedule->>GetBatch: get_batch_on_this_rank_for_sequence_packing
        Note over GetBatch: Broadcast tokens/labels to TP group
        Note over GetBatch: Partition for CP if needed
        GetBatch-->>Schedule: (tokens, labels, loss_mask, pos_ids, packed_seq_params)
    end
    
    Schedule-->>Train: forward_data_store + [total_tokens, seq_sq_sum]
```

### New Arguments

| Argument | Type | Description |
|----------|------|-------------|
| `--sequence-packing` | flag | Enable sequence packing (THD format) for training |
| `--sequence-packing-scheduler` | str | Scheduler type: `default` or `empty` |
| `--sft-mock-dataset-config-json` | str | JSON config for mock dataset |

### Changes

| File | Description |
|------|-------------|
| `megatron/core/datasets/data_schedule.py` | Core scheduling and packing logic |
| `megatron/core/pipeline_parallel/schedules.py` | Integration with forward/backward schedules |
| `megatron/training/training.py` | Updated FLOPs calculation for variable-length sequences |
| `megatron/training/datasets/sft_dataset.py` | Mock dataset support |
| `megatron/training/arguments.py` | New CLI arguments |
| `megatron/core/model_parallel_config.py` | Configuration options |
| `tests/unit_tests/test_sequence_packing.py` | Unit tests |

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Phlip79** [comment]
**We are changing our review process and marking all open, unlabeled PRs as draft. This change will go in effect starting once #3659 is merged.**

Moving forward, all PRs will be required to start as draft PRs. If you wish to get your PR merged, mark your PR as “Ready for review”. Read more about the new process at [submit.md](https://github.com/Phlip79/Megatron-LM/blob/d3763629bcdd04acca52477af7eb8795090426ff/docs/developer/submit.md).

**Review by cuichenx — APPROVED** [review]

File: pretrain_gpt.py [hunk 1]
```diff
@@ -318,6 +329,8 @@ def core_gpt_dataset_config_from_args(args: Any) -> GPTDatasetConfig:
         "data_parallel_size": args.data_parallel_size,
         "sequence_parallel_size": args.tensor_model_parallel_size * args.sequence_parallel,
         "hybrid_context_parallel": args.hybrid_context_parallel,
+        "sft_mock_dataset_config_json":args.sft_mock_dataset_config_json,
```
Comment (jkamalu) [inline pretrain_gpt.py]: formatting

File: megatron/training/datasets/sft_dataset.py [hunk 2]
```diff
@@ -190,3 +213,158 @@ def extend_with_padding(tokens, targets, positions, pad_len):
...
+        return self.size
+
+    def __getitem__(self, idx: int) -> List[np.ndarray]:
+        # the length of sample is 'length', but only length-1 elements are generated here, 
+        # because an eod token will be appended at the end later in SFTDataset
+
+        length = self.sequence_lengths[idx % self.size]
+        sample = np.arange(1, length, dtype=np.int64)
+        return sample
+class MockSFTDataset(SFTDataset):
```
Comment (jkamalu) [inline megatron/training/datasets/sft_dataset.py:274]: formatting.

**Comment by Phlip79** [comment]
@xiaoyao0115 it would be helpful with the merge process if you please split this PR up. I recommend using the `split-pr` skill. There are also unresolved comments from Claude, @jkamalu, and conflicts with main.

File: pretrain_gpt.py (hunk 1 above)
Comment (xiaoyao0115) [reply pretrain_gpt.py]: fixed

File: megatron/training/datasets/sft_dataset.py [hunk 3]
```diff
@@ -2,9 +2,12 @@
 
 import atexit, json
 from collections import Counter
-from typing import Any, Dict, Optional
+import json
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/training/datasets/sft_dataset.py]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        for gid in send_ids_sorted:
+            t = batch[gid2local_id[gid]][key].to(torch.cuda.current_device(), non_blocking=True)
+            flattened_tensors.append(t.reshape(-1))
+        return (
+            torch.cat(flattened_tensors, dim=0)
+            if flattened_tensors
+            else torch.empty(1, device=torch.cuda.current_device(), dtype=batch[0][key].dtype)
+        )
+
+    def _unpack_sample_by_key(key: str, recv_tensor: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:436]: fixed

File: megatron/training/arguments.py [hunk 5]
```diff
@@ -2741,6 +2760,14 @@ def _add_distributed_args(parser):
...
                        'specify separated types for each layer like '
                        '--cp-comm-type p2p p2p a2a a2a a2a+p2p a2a+p2p')
+    group.add_argument('--max-seqlen-per-dp-cp-rank', type=int, default=None,
+                       help='Maximum sequence length per CP rank. This is used to calculate the '
+                       'number of sub-samples assigned to each CP rank when using heterogeneous context parallel.')
+    group.add_argument('--hybrid-context-parallel', action='store_true', default=False,
+                       help='Enables hybrid context parallel. This is used to balance the workload '
+                       'of each CP rank when we use packed samples with variable sequence lengths. '
+                       'Requires --max-seqlen-per-dp-cp-rank to be set.')
+    group.add_argument('--sequence-packing-scheduler', type=str, default='default_sequence_packing', choices=['default_sequence_packing'])
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/training/arguments.py]: fixed

File: megatron/training/datasets/sft_dataset.py (hunk 2 above)
Comment (xiaoyao0115) [reply megatron/training/datasets/sft_dataset.py:274]: fixed

File: megatron/core/datasets/data_schedule.py [hunk 6]
```diff
@@ -299,3 +352,584 @@ def __next__(self) -> Any:
...
+
+    return (
+        new_data_iterator,
+        num_micro_batches,
+        seqlen_sum_this_global_batch,
+        seqlen_squared_sum_this_global_batch,
+    )
+
+
+def get_batch_on_this_rank_for_sequence_packing(
```
Comment (parthmannan) [inline megatron/core/datasets/data_schedule.py:744]: SFT/THD format support added a similar function [here](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/utils.py#L2256). Worth keeping the `get_batch_*` related function in `utils.py` and consolidating any necessary changes there.

File: megatron/core/datasets/gpt_dataset.py
Comment (jkamalu) [inline megatron/core/datasets/gpt_dataset.py:1]: Hm, we shouldn't be putting configs within configs. We shouldn't be passing around configs as json strings. Is there an alternative here?

File: megatron/core/datasets/data_schedule.py [hunk 7]
```diff
@@ -299,3 +352,584 @@ def __next__(self) -> Any:
...
+            if pp_group.rank() == 0:
+                # the first stage
+                data_iterator = data_iterator[0]
+            elif pp_group.rank() == pp_group.size() - 1:
+                # the last stage
+                data_iterator = data_iterator[-1]
+            else:
+                data_iterator = None
+
+        # data_iterator is not None when TP rank 0, with PP stage 0 or -1.
```
Comment (guihong-nv) [inline megatron/core/datasets/data_schedule.py:554]: Packed datasets build an iterator on every TP-rank-0 PP stage, but the scheduler permits iterators only on the first/last stages. Any non-VPP run with PP ≥ 3 therefore aborts on a middle stage before its first forward pass. The PP=4 test masks this by manually assigning `None` to middle-stage iterators.

File: megatron/core/datasets/data_schedule.py [hunk 8]
```diff
@@ -299,3 +352,584 @@ def __next__(self) -> Any:
...
+            batch['loss_mask'] = batch['loss_mask'].view(1, total_tokens)
+        else:
+            batch['labels'] = torch.empty([1, total_tokens], dtype=torch.int64, device=dev)
+            batch['loss_mask'] = torch.empty([1, total_tokens], dtype=torch.float32, device=dev)
+    else:
+        # Non last stage rank doesn't need labels and loss_mask.
+        batch['labels'] = None
+        batch['loss_mask'] = None
+
+    # Step3: Prepare "padding_mask" on all TP ranks.
```
Comment (guihong-nv) [inline megatron/core/datasets/data_schedule.py:876]: Every stage receives a full CP-local padding mask. GPTModel TP-shards it only on the embedding stage; downstream PP/VP stages retain S mask entries while sequence-parallel activations contain S/TP tokens. With PP > 1, TP > 1, sequence parallelism, and downstream MoE layers, router masking encounters incompatible dimensions.

File: megatron/core/datasets/data_schedule.py [hunk 9]
```diff
@@ -299,3 +352,584 @@ def __next__(self) -> Any:
...
+        num_packed_sequence = len(packed_id_groups)
+
+        # when enabling vpp, we want the number of packed sequences to be
+        # multiple of dp_size * microbatch_group_size_per_vp_stage
+        multiple = self.dp_size * (
+            self.microbatch_group_size_per_vp_stage
+            if self.microbatch_group_size_per_vp_stage is not None
+            else 1
+        )
+        if num_packed_sequence % multiple != 0:
```
Comment (guihong-nv) [inline megatron/core/datasets/data_schedule.py:471]: The loop asserts `i > 0` before inspecting group zero. For a valid `DP=1, PP=2, VPP=2` batch where two short samples fit into one packed group, alignment needs to split that group into two but immediately raises `Not enough samples to move`. Group zero should remain splittable while it contains more than one sample.

**Comment by ilml** [comment]
sync with team offline on Jun 30th, conclusion:
1. break down into 2 prs, @ilml and @xiaoyao0115 will drive it.
2. api redesign, code cleanup, dedup.


Current data iterator wrapper - [Megatron-LM/megatron/core/datasets/data_schedule.py at main · NVIDIA/Megatron-LM](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/data_schedule.py#L12)
 
Current scheduler - [Megatron-LM/megatron/core/pipeline_parallel/hybrid_cp_schedule.py at main · NVIDIA/Megatron-LM](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/pipeline_parallel/hybrid_cp_schedule.py#L14)
 
Adding packing to the current scheduler - [Add packing support to Dynamic CP Schedule by parthmannan · Pull Request #5191 · NVIDIA/Megatron-LM](https://github.com/NVIDIA/Megatron-LM/pull/5191/changes)

https://github.com/NVIDIA/Megatron-LM/pull/5191 is just scheduler changes 

@zhongbozhu @deepakn94 @parthmannan @yaoyu-33 @cuichenx @YangFei1990 @liding-nv  @xiaoyao0115

**Comment by FDecaYed** [comment]
Recording the replacement plan for the DeepSeek-V4 HybridModel recut: this historical PR is no longer an active merge path. The authoritative guide is [DeepSeek-V4 HybridModel merge plan #5795](https://github.com/NVIDIA/Megatron-LM/pull/5795).

The DSv4-scoped work from this PR is being reconstructed as:
- **P7 / #5993** — fixed DP-balanced THD packing scheduler
- **P8** — HybridModel-only variable-length and mock dataset support
- **P11** — static-shape THD Transformer Engine CUDA graphs
- **D9** — Hybrid-only FLOP accounting, functional E2E coverage, and the final 64K recipe

GPT/SFT-specific and dynamic-context-parallel extensions are outside the frozen HybridModel-only scope. The original work remains linked from #5795 for provenance and credit. This PR will stay open as a draft reference, and the `deepseekv4` label is being removed so it is not tracked as an active delivery PR.
